### PR TITLE
Lua API: Add more inventory callbacks

### PR DESCRIPTION
**Pull request by SmallJoker**
Rebase & conflict resolved version of #4035
The only thing that effectively changed are the docs - they should be easier to understand now.

### CODE REVIEW

**Comment by sfan5** [comment]
`core.registered_on_player_inventory_*` ok
`core.registered_on_nodemeta_inventory_*` would be better if these were in the nodedef
`core.registered_on_detached_inventory_*` i don't see any usecase for this

**Comment by raymoo** [comment]
Is it possible to modify/cancel certain inventory operation like it is with existing callbacks?

**Comment by nerzhul** [comment]
it only build on windows and satisfy LINT. Please fix

**Comment by SmallJoker** [comment]
@nerzhul I bundled the two scripting sides together into a new class `Scripting` for adding CSM callbacks in the future. The main problem that made Travis fail is solved now there's still a pointer that isn't set properly (see commit description).
I'm quite out of ideas of how to implement the scripting in a way that won't cause a huge work if it's going to be added to CSM aswell. But passing the scripting pointer to each function like done in the original pull doesn't look nice and can surely be simplified somehow. If you've got an idea about implementing this - I'd be glad to hear it.

File: src/inventory.h [hunk 1]
```diff
@@ -232,28 +235,78 @@ class InventoryList
...
+	InventoryChangeReceiver *m_rec;
+};
+
+class InventoryChangeReceiver
+{
+public:
+	InventoryChangeReceiver() { }
+	InventoryChangeReceiver(const Scripting script)
+	{
+		m_script = script;
```
Comment (nerzhul) [inline src/inventory.h]: Ouch, you copy all Scripting interface ? it should be a pointer

**Comment by nerzhul** [comment]
You can already access to client and server from ModAPIBase, what you need is a module like l_player dedicated to inventories for the callbacks, why not re-use simple l_inventory.cpp and add you callback there, it's really a good place

File: src/inventory.h (hunk 1 above)
Comment (rubenwardy) [inline src/inventory.h]: copying should be disabled for Scripting, anyway (ie: this shouldn't even compile)

**Comment by nerzhul** [comment]
I can provide a trivial patch to ensure this will never work

**Comment by SmallJoker** [comment]
@nerzhul I've planned to work on this further this weekend.
> to ensure this will never work

Do you mean the current state of this pull or the target functionality? However, I appreciate if you have patches :)

**Comment by nerzhul** [comment]
I pushed this morning to master a patch to forbid class copy for the three Scripting interfaces

**Comment by paramat** [comment]
Seems complex, maybe over complex? (like the author's other PR), can it be simplified to the more useful features?

**Comment by SmallJoker** [comment]
R.I.P. will need an entire redo, as 1) the scripting part needs reworking and 2) this inventory callback linking should happen in a less complex way.

**Pull request closed without merging**